### ISSUE TITLE
Change input type of readers so source positions can be tracked

### ISSUE THREAD

**Issue by jgm** [jgm/pandoc#7249]
Currently the readers don't know which file a given part of the input text comes from, since all the inputs are concatenated into one Text before being passed to the reader.

Proposal: generalize reader types from
```
PandocMonad m => ReaderOptions -> Text -> m Pandoc 
```
to
``` hs
(PandocMonad m, PandocInput t) => ReaderOptions -> t -> m Pandoc 
```

A typeclass would give good backwards compatibility, allowing current invocations of the readers which assume a Text parameter to continue to work without changes.

But we could also define instances like `[(FilePath, Text)]` which would allow the parser to update source position accurately and resolve paths relative to the particular file in which a reference occurs (see #5501).  Or something like `data HttpInput = HttpInput Encoding URL Text` for content from URL parameters.

It would be relatively straightforward to provide Stream instances for parsec parsers that would update source position when we move to text from a different file.  (The one complication I see, possibly a serious one, is with parsers that use `getInput`.)  The tagsoup tokenizer we use with HTML content should also be able to make use of this.  I'm not sure about the parser we use for XML formats, though.

Previously discussed at  <https://groups.google.com/g/pandoc-discuss/c/M_UPUFs1G6o/m/hKGN-V8YBwAJ> and
<https://github.com/jgm/pandoc/issues/5501#issuecomment-809607817>.

**Comment by jgm** [comment]
Current plan:

- [ ] Add `newtype Sources = Sources { unSources :: [(SourcePos, Text)] }`
- [ ] Add instance for `Stream Sources m Char`
- [ ] Replace `ParsecT Text` with `ParsecT Sources` wherever we have it
- [ ] Add `ToSources` typeclass, with method `toSources :: a -> Sources`
- [ ] Add `ToSources` instances for `Text` and `Sources`.  (*Alternative:* If we wanted to get fancy, we could make the type of `toSources :: PandocMonad m => a -> m Sources`; this would allow us to create a `ToSources` instance for `FilePath`, `URL`, etc...is this a good idea?  We don't really lose much by making the type more general in this way, since we only use `toSources` at the beginning of each reader, and the readers are all in PandocMonad.)
- [ ] Change text readers to take an instance of `ToSources` rather than `Text` as parameter.  Use `toSources` on the input to change it into a `Sources`to pass to the parser.
- [ ] Update tests and benchmark.
- [ ] Update `doc/using-the-pandoc-api.md`

@tarleb any thoughts?

**Comment by jgm** [comment]
Note that handling include files and doing "parse from string" will be much simpler with this interface.  Just insert a `(SourcePos, Text)` at the beginning of the input.

**Comment by jgm** [comment]
OK, this doesn't work, because the `uncons` method of `Stream` doesn't allow you to adjust source position.

**Comment by jgm** [comment]
I'm not sure I see a way forward with this, short of doing what we do for commonmark -- tokenizing the input into a stream of tokens, each marked with source position.  (But this would require rewriting every parser.)

Since the commonmark parser already does this, we could try to set things up so that at least *it* has access to accurate source positions.  This could be done by exposing a reader with a different signature (e.g. `[(FilePath, Text)]` or `Either Text [FilePath]`) and special-casing it in App.  The same could be done with the LaTeX reader, which also uses tokens with source positions, and perhaps also for HTML.

**Comment by jgm** [comment]
I figured out a way to do this, I think -- using megaparsec rather than parsec. Proof of concept:

``` haskell
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeFamilies #-}
import Text.Megaparsec
import Text.Megaparsec.Char
import Data.Text
import qualified Data.List.NonEmpty as NonEmpty
import Data.Proxy
import Data.Void
import qualified Data.Text as T

newtype Sources = Sources [(FilePath, Text)]
  deriving (Show)

instance Stream Sources where
  type Token Sources = Char
  type Tokens Sources = Text

  tokensToChunk _ = T.pack

  chunkToTokens _ = T.unpack

  chunkLength _ = T.length

  take1_ (Sources []) = Nothing
  take1_ (Sources ((fp,t):rest)) =
    case T.uncons t of
      Nothing -> Nothing
      Just (c,t') -> Just (c, if T.null t'
                                 then Sources rest
                                 else Sources ((fp,t'):rest))

  takeN_ _ (Sources []) = Nothing
  takeN_ n (Sources ((fp,t):rest)) =
    case T.splitAt n t of
      (x,y) ->
        let len = T.length x
         in if len < n
               then (\(x',y') -> (x <> x', y'))
                       <$> takeN_ (n - len) (Sources rest)
               else Just (x, if T.null y
                                then Sources rest
                                else Sources ((fp,y):rest))

  takeWhile_ f (Sources []) = (mempty, Sources [])
  takeWhile_ f (Sources ((fp,t):rest)) =
    case T.span f t of
      (x,y) ->
         if T.null y
            then (\(x',y') -> (x <> x', y'))
                $ takeWhile_ f (Sources rest)
            else (x, Sources ((fp,y):rest))

instance VisualStream Sources where
  showTokens _prx = NonEmpty.toList
  tokensLength _prx = NonEmpty.length


instance TraversableStream Sources where

  reachOffset off ps
    = let Sources xs = pstateInput ps
          go _ ps' [] = (Nothing, ps')
          go o ps' ((fp,t):rest) =
            if o > T.length t
               then
                 case rest of
                   [] -> (Nothing, ps') -- is this okay?
                   (fp',t'):_ -> go (o - T.length t)
                                   ps'{ pstateInput = Sources rest
                                      , pstateOffset = o - T.length t
                                      , pstateSourcePos = initialPos fp' }
                                   rest
                else
                  let (x,y) = T.splitAt o t
                      curpos = pstateSourcePos ps'
                      nls = T.count (T.singleton '\n') x
                      tw = unPos (pstateTabWidth ps)
                      updateCol _   '\n' = 0
                      updateCol col '\t' =
                         col + (tw - ((col - 1) `mod` tw))
                      updateCol col _ = col + 1
                      s = T.unpack
                          (T.takeWhileEnd (/='\n') x <>
                            T.takeWhile (/='\n') y)

                   in (Just s,
                       ps{ pstateInput = Sources ((fp, y):rest)
                         , pstateOffset = off
                         , pstateSourcePos =
                            SourcePos
                              { sourceName = fp
                              , sourceLine =
                                  mkPos $ unPos (sourceLine curpos) + nls
                              , sourceColumn = mkPos $
                                  T.foldl' updateCol
                                    (unPos (sourceColumn curpos))
                                    (T.takeWhileEnd (/='\n') x)
                              }
                         })


       in case off - pstateOffset ps of
            o | o < 0 -> (Nothing, ps) -- Is this right?
              | otherwise -> go o ps xs


test :: IO ()
test = do
  parseTest myParser (Sources [("a","aaaaa\n"),("b","bbbb\nb2b2b2\n")])

myParser :: Parsec Void Sources Int
myParser = do
  (sum <$> many (1 <$ satisfy (/='2')) <* eof)
```

**Comment by jgm** [comment]
Looks like include files might not work with this approach.  IF you use `setInput` to change the input, the error message at the end is computed based on the original input.
```hs
test2 :: IO ()
test2 = do
  parseTest (myParser2 >> myParser)
    (Sources [("a","aaaaa\n"),("b","bbbb\nb2b2b2\n")])

myParser2 :: Parsec Void Sources ()
myParser2 = do
  Sources xs <- getInput
  setInput $ Sources $ ("c","ccc3ccc\n")
```
output:
```
*Main> test2
b:1:1:
unexpected 2
expecting end of input
```

**Comment by jgm** [comment]
I think I figured out how to do this with parsec.  The trick is that you have to redefine `satisfy` so that it takes the Sources stream into account in calculating source position. Example:
```hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
import Text.Parsec hiding (satisfy)
import Text.Parsec.Pos
import Data.Text (Text)
import qualified Data.Text as T
import Control.Monad.Identity

newtype Sources = Sources [(FilePath,Text)]

instance Monad m => Stream Sources m Char where
  uncons (Sources []) = return Nothing
  uncons sources@(Sources ((fp,t):rest)) =
    case T.uncons t of
      Nothing -> uncons (Sources rest)
      Just (c,t') -> return $ Just (c, Sources ((fp,t'):rest))

satisfy :: Monad m => (Char -> Bool) -> ParsecT Sources u m Char
satisfy f = tokenPrim pprint nextpos matcher
 where
  pprint = show
  nextpos pos c sources =
    case sources of
      Sources [] -> updatePosChar pos c
      Sources ((fp,t):_)
        | T.null t -> initialPos fp
        | otherwise -> setSourceName (updateCharPos pos c) fp
  matcher c = if f c then Just c else Nothing
  updateCharPos pos c
    = case c of
        '\n' -> incSourceLine (setSourceColumn pos 1) 1
        '\t' -> incSourceColumn pos (4 - ((sourceColumn pos - 1) `mod` 4))
        _    -> incSourceColumn pos 1

p :: ParsecT Sources () Identity Int
p = (sum <$> many (1 <$ satisfy (/='x'))) <* eof

test :: IO ()
test = do
  t1 <- T.pack <$> readFile "a.txt"
  t2 <- T.pack <$> readFile "b.txt"
  print $ runParserT p () "" (Sources [("a.txt",t1),("b.txt",t2)])
```

**Comment by jgm** [comment]
Preview now available in `sources` branch.

**Comment by jgm** [comment]
I have run benchmarks and these changes do not affect performance.

**Comment by jgm** [comment]
Closed by 6e45607f9948f45b2e94f54b4825b667ca0d5441
